src/snowflake/sqlalchemy/base.py

```diff
@@ -150,6 +152,163 @@ def _split_schema_by_dot(self, schema):
+
+        text += "INTO "
+        table_text = preparer.format_table(insert_stmt.table)
+
+        if insert_stmt._hints:
+            _, table_text = self._setup_crud_hints(insert_stmt, table_text)
+
+        if insert_stmt._independent_ctes:
+            for cte in insert_stmt._independent_ctes:
+                cte._compiler_dispatch(self, **kw)
+
+        text += table_text
+
+        if crud_params_single or not supports_default_values:
+            text += " (%s)" % ", ".join([expr for c, expr, value in crud_params_single])
+
+        if self.returning or insert_stmt._returning:
+            returning_clause = self.returning_clause(
+                insert_stmt, self.returning or insert_stmt._returning
+            )
+
+            if self.returning_precedes_values:
+                text += " " + returning_clause
+        else:
+            returning_clause = None
+
+        if insert_stmt.select is not None:
+            # placed here by crud.py
+            select_text = self.process(
+                self.stack[-1]["insert_from_select"], insert_into=True, **kw
+            )
+
+            if self.ctes and self.dialect.cte_follows_insert:
+                nesting_level = len(self.stack) if not toplevel else None
+                text += " {}{}".format(
+                    self._render_cte_clause(
+                        nesting_level=nesting_level,
+                        include_following_stack=True,
+                    ),
+                    select_text,
+                )
+            else:
+                text += " %s" % select_text
+        elif not crud_params and supports_default_values:
+            text += " DEFAULT VALUES"
+        elif compile_state._has_multi_parameters:
+            text += " VALUES %s" % (
+                ", ".join(
+                    "(%s)" % (", ".join(value for c, expr, value in crud_param_set))
+                    for crud_param_set in crud_params
+                )
+            )
+        else:
+            crud_params_contain_array_type = any(
+                [
+                    crud_column
+                    for crud_column in crud_params
+                    if isinstance(crud_column[0].type, ARRAY)
+                ]
+            )
+            if crud_params_contain_array_type:
+                # for ARRAY type, use select style insert
+                # insert_single_values_expr only applies to VALUES expression according to sqlalchemy definition
+                # thus setting it to None
+                insert_single_values_expr = None
+                select_expression = ", ".join(
+                    [
+                        ("ARRAY_CONSTRUCT(%s)" % value)
+                        if isinstance(c.type, ARRAY)
+                        else ("%s" % value)
+                        for c, expr, value in crud_params
+                    ]
+                )
+                text += " SELECT %s" % select_expression
```

Comment:
this is the implementation for array data insertion, all other code in this `visit_insert` are just copied pasted from sql alchemy.